service/history/decisionHandler.go

```diff
@@ -654,24 +662,66 @@ func (handler *decisionHandlerImpl) handleBufferedQueries(
 		buffered := queryRegistry.getBufferedIDs()
 		for _, id := range buffered {
 			if err := queryRegistry.setTerminationState(id, failedTerminationState); err != nil {
-				handler.logger.Error("failed to fail query", tag.QueryID(id), tag.Error(err))
-				handler.metricsClient.IncCounter(metrics.HistoryRespondDecisionTaskCompletedScope, metrics.FailQueryFailedCount)
+				handler.logger.Error(
+					"failed to set query termination state to failed",
+					tag.WorkflowDomainName(domain),
+					tag.WorkflowID(workflowID),
+					tag.WorkflowRunID(runID),
+					tag.QueryID(id),
+					tag.Error(err))
+				scope.IncCounter(metrics.FailQueryFailedCount)
 			}
 		}
 		return
 	}
 
-	// Complete all queries for which answers have been received.
+	sizeLimitError := handler.config.BlobSizeLimitError(domain)
+	sizeLimitWarn := handler.config.BlobSizeLimitWarn(domain)
+
+	// Complete or fail all queries we have results for
 	for id, result := range queryResults {
-		completeTerminationState := &queryTerminationState{
-			queryTerminationType: queryTerminationTypeCompleted,
-			queryResult:          result,
-		}
-		if err := queryRegistry.setTerminationState(id, completeTerminationState); err != nil {
-			handler.logger.Error("failed to complete query", tag.QueryID(id), tag.Error(err))
-			handler.metricsClient.IncCounter(metrics.HistoryRespondDecisionTaskCompletedScope, metrics.CompleteQueryFailedCount)
+		if err := common.CheckEventBlobSizeLimit(
+			len(result.GetAnswer()),
+			sizeLimitWarn,
+			sizeLimitError,
+			domainID,
+			workflowID,
+			runID,
+			scope,
+			handler.throttledLogger,
+		); err != nil {
+			failedTerminationState := &queryTerminationState{
+				queryTerminationType: queryTerminationTypeFailed,
+				failure:              err,
+			}
+			if err := queryRegistry.setTerminationState(id, failedTerminationState); err != nil {
+				handler.logger.Error(
+					"failed to set query termination state to failed",
+					tag.WorkflowDomainName(domain),
+					tag.WorkflowID(workflowID),
+					tag.WorkflowRunID(runID),
+					tag.QueryID(id),
+					tag.Error(err))
+				scope.IncCounter(metrics.FailQueryFailedCount)
+			}
```

Comment:
nit: `continue` here and we don't need the `else` below? For me it's a little bit more readable.